Intersections_2/include/CGAL/Intersections_2/Circle_2_Segment_2.h

```diff
@@ -13,32 +13,109 @@
 #ifndef CGAL_INTERSECTIONS_2_CIRCLE_2_SEGMENT_2_H
 #define CGAL_INTERSECTIONS_2_CIRCLE_2_SEGMENT_2_H
 
-#include <CGAL/Distance_2/Point_2_Segment_2.h>
+#include <CGAL/Distance_2/internal/squared_distance_utils_2.h>
 #include <CGAL/Intersection_traits_2.h>
+#include <CGAL/tags.h>
 
 #include <CGAL/Circle_2.h>
 #include <CGAL/Segment_2.h>
 
 namespace CGAL {
 namespace Intersections {
 namespace internal {
+//
+
+template <class K>
+typename K::Boolean
+do_intersect(const typename K::Circle_2& circ,
+             const typename K::Segment_2& s,
+             const K& k,
+             const Homogeneous_tag&)
+{
+  typedef typename K::Vector_2 Vector_2;
+  typedef typename K::Point_2 Point_2;
+  typedef typename K::FT FT;
+
+  typename K::Construct_vector_2 vector = k.construct_vector_2_object();
+  typename K::Compute_squared_distance_2 sq_dist = k.compute_squared_distance_2_object();
+
+  const Point_2 c = circ.center();
+  const FT sqrad = circ.squared_radius();
+  const Vector_2 diffsrc = vector(s.source(), c);
+  const Vector_2 difftgt = vector(s.target(), c);
+
+  FT lsrc = CGAL::internal::wdot(diffsrc, diffsrc, k);
+  FT lsrcw = square(diffsrc.hw());
+  FT ltgt = CGAL::internal::wdot(difftgt, difftgt, k);
+  FT ltgtw = square(difftgt.hw());
+
+  if (lsrc < sqrad * lsrcw) {
+    if (ltgt < sqrad * ltgtw)
+      return false;
+    else
+      return true;
+  }
+  else if (ltgt <= sqrad * ltgtw)
+    return true;
+
+  return sq_dist(c, s) <= sqrad;
+}
+
+template <class K>
+typename K::Boolean
+do_intersect(const typename K::Circle_2& circ,
+             const typename K::Segment_2& s,
+             const K& k,
+             const Cartesian_tag&)
+{
+  typedef typename K::Vector_2 Vector_2;
+  typedef typename K::Point_2 Point_2;
+  typedef typename K::FT FT;
+
+  typename K::Construct_vector_2 vector = k.construct_vector_2_object();
+  typename K::Compute_squared_length_2 sq_length = k.compute_squared_length_2_object();
+  typename K::Compute_squared_distance_2 sq_dist = k.compute_squared_distance_2_object();
+
+  const Point_2 c = circ.center();
+  const FT sqrad = circ.squared_radius();
+  const Vector_2 diffsrc = vector(s.source(), c);
+  const Vector_2 difftgt = vector(s.target(), c);
+
+  FT lsrc = sq_length(diffsrc);
+  FT ltgt = sq_length(difftgt);
+
+  if (lsrc < sqrad) {
+    if (ltgt < sqrad)
+      return false;
+    else
+      return true;
+  }
+  else if (ltgt <= sqrad)
+    return true;
+
+  return sq_dist(c, s) <= sqrad;
```

Comment:
here we already know that the orthogonal projection of the center is within src and tgt so we can use the line (to not redo the tests)